openstack/02a-MAClow/IEEE802154E.c

```diff
@@ -1732,13 +1729,24 @@ port_INLINE void asnStoreFromAdv(uint8_t* asn) {
    ieee154e_vars.asn.bytes2and3   =     asn[2]+
                                     256*asn[3];
    ieee154e_vars.asn.byte4        =     asn[4];
-   
    // determine the current slotOffset
-   /*
-   Note: this is a bit of a hack. Normally, slotOffset=ASN%slotlength. But since
-   the ADV is exchanged in slot 0, we know that we're currently at slotOffset==0
-   */
-   ieee154e_vars.slotOffset       = 0;
+}
+
+port_INLINE void ieee154e_syncSlotOffset() {
+   frameLength_t frameLength;
+   uint32_t slotOffset;
+   
+   frameLength = schedule_getFrameLength();
+   slotOffset = ieee154e_vars.asn.byte4;
+   slotOffset = slotOffset % frameLength;
+   slotOffset = slotOffset << 16;
+   slotOffset = slotOffset + ieee154e_vars.asn.bytes2and3;
+   slotOffset = slotOffset % frameLength;
+   slotOffset = slotOffset << 16;
+   slotOffset = slotOffset + ieee154e_vars.asn.bytes0and1;
+   slotOffset = slotOffset % frameLength;

```

Comment:
The counts in the example start from considering a 4 byte number:
0x01010101 % 5 = 4 TRUE (this is a 4 byte number)
0x01 % 5 = 1 TRUE (first significant byte)
0x10101 % 5 = 3 TRUE (shift of the result of 2 bytes left and sum with the following 2 bytes)
but now the shift to be applied should be of 8 bits, because we are going to sum up only the 4th byte:
0x301 % 5 = 4

If we consider a 5 bytes number:
0x0101010101 % 5 = 0
0x01 % 5 = 1
(0x010000 + 0x0101) % 5 = 0x010101 % 5 = 3
(0x030000 + 0x0101) % 5 = 0x030101 % 5 = 0